proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -287,130 +267,232 @@ protocol TargetBuildContext {
+        /// Optional initial working directory of the command.
+        workingDirectory: Path? = nil,
+        /// Any custom settings for the process environment.
+        environment: [String: String] = [:],
+        /// A directory into which the command can write output files that
+        /// should be processed further.
+        outputFilesDirectory: Path
+    )
 }
 
 /// Emits errors, warnings, and remarks to be shown as a result of running
 /// the plugin. If any errors are emitted, the plugin is considered to have
 /// have failed, which will be reported to users during the build.
 protocol DiagnosticsEmitter {
-    func emit(error message: String, file: FilePath? = nil, line: Int? = nil)
-    func emit(warning message: String, file: FilePath? = nil, line: Int? = nil)
-    func emit(remark message: String, file: FilePath? = nil, line: Int? = nil)
+    func emit(error message: String, file: Path? = nil, line: Int? = nil)
+    func emit(warning message: String, file: Path? = nil, line: Int? = nil)
+    func emit(remark message: String, file: Path? = nil, line: Int? = nil)
+}
+
+/// Provides information about a list of files. The order is not defined
+/// but is guaranteed to be stable. This allows the implementation to be
+/// more efficient than a static file list.
+protocol FileList: Sequence {
+    func makeIterator() -> FileListIterator
+}
+struct FileListIterator: IteratorProtocol {
+    mutating func next() -> FileInfo?
+}
+
+/// Provides information about a single file in a FileList.
+protocol FileInfo {
+    var path: Path { get }
+    var type: FileType { get }
+}
+
+/// Provides information about a the type of a file. Any future cases will
+/// use availability annotations to make sure existing plugins still work
+/// until they increase their required tools version.
+enum FileType {
+    /// A source file of a kind accepted by the Swift compiler. Only
+    /// present in Swift target.
+    case swiftSourceFile
+    /// A source file of a kind accepted by the Clang compiler. Only
+    /// present in Clang targets.
+    case clangSourceFile
+    /// A resource file of any kind either supported natively by SwiftPM
+    /// or declared as a resource by the target.
+    case resourceFile
+    /// A file not covered by any other rule.
+    case unknownFile
 }
 
-/// In addition, the API of `FilePath` from SwiftSystem will be available.
+/// A simple representation of a path in the file system.
+protocol Path {
+    /// The last path component (including any extension).
+    public var filename: String { get }
+    /// The last path component (without any extension).
+    public var stem: String { get }
+    /// The filename extension, if any.
+    public var `extension`: String? { get }
+    /// The path except for the last path component.
+    public var parentDirectory: String { get }
+    /// The result of appending one or more path components.
+    public func appending(_ other: String, ...) -> Path
+  }
 ```
```

Comment:
how does that Path relate to what Swift System offers? Is System going to be an implementation detail?

Reply:
I see this is answered in `A More Sophisticated `Path` Type` :)